### PR TITLE
fix(homepage): a11y errors

### PR DESCRIPTION
**Pull request by 0618**
#### Description of changes

| before | after |
| --- | --- |
| Screen Shot 2022-03-02 at 16 18 32 |  ~~`image`~~ image |
| 6 X Empty form label | ✅  |
| 1 X Language missing or invalid | ✅  |
| 6 X Empty button | ✅  |
| 5 X Broken ARIA reference |  ✅  |
| image | image |
| image |  image |

**Other code change**

- fix type error
- rename variables
- change format. double quote to single quote.

**Note**

There are still some a11y alerts on the homepage and I didn't fix them because:

- **1 X Missing first level heading**:  I'm not sure what to use as `h1`.
- **Redundant link**: both links on the bottom are "https://docs.amplify.aws/console/"
- **2 X Noscript element**: those are empty tags

**TODO**:
- [x] There are several primitives/components related a11y errors are converted in https://github.com/aws-amplify/amplify-ui/pull/1483/commits/dde5608889d4d22e095a97d70a2101ce9c50a145 and will be done in another PR https://github.com/aws-amplify/amplify-ui/pull/1495

#### Issue #, if available

https://app.asana.com/0/1201736086077838/1201759796183705/f

#### Description of how you validated changes

- ran `yarn test`

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are updated
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/src/pages/index.page.tsx [hunk 1]
```diff
@@ -170,16 +171,17 @@ const HomePage = ({ colorMode, setThemeOverride, themeOverride }) => {
...
-                label=""
-                labelHidden={true}
-                isReadOnly={true}
-                className="install-code"
-                outerEndComponent={
-                  <Copy variation="link" text={frameworkInstallScript} />
-                }
-                value={frameworkInstallScript}
-              />
+              <code className="install-code__container">
```
Comment (0618) [inline docs/src/pages/index.page.tsx:174]: `<code>` tag might be more proper here

File: docs/src/components/Copy.tsx [hunk 2]
```diff
@@ -15,7 +15,12 @@ export const Copy = ({ text, size, variation }) => {
 
   return (
     <CopyToClipboard text={text} onCopy={copy}>
-      <Button size={size} variation={variation} isLoading={copied}>
+      <Button
+        size={size}
+        variation={variation}
+        isLoading={copied}
+        className={className}
```
Comment (0618) [inline docs/src/components/Copy.tsx:22]: add `className` because of the modification on `docs/src/pages/index.page.tsx`

File: docs/src/components/Layout/Header.tsx [hunk 3]
```diff
@@ -113,6 +113,7 @@ export const Header = ({ platform, colorMode, setColorMode }) => {
         <Button
           className="docs-header-menu-button"
           onClick={() => setExpanded(!expanded)}
+          aria-label="Docs header menu button"
```
Comment (0618) [inline docs/src/components/Layout/Header.tsx:116]: we may eventually use variables for aria-label for i18n.

File: docs/src/pages/_document.page.tsx [hunk 4]
```diff
@@ -45,7 +45,7 @@ const getCSPContent = (context: Readonly<HtmlProps>) => {
 class MyDocument extends Document {
   render() {
     return (
-      <Html>
+      <Html lang="en-us">
```
Comment (0618) [inline docs/src/pages/_document.page.tsx:48]: we may eventually use variables for lang

File: packages/react/src/primitives/Menu/Menu.tsx [hunk 5]
```diff
@@ -33,7 +33,7 @@ const MenuPrimitive: Primitive<MenuProps, 'div'> = (
     <DropdownMenuTrigger asChild={true}>
       {trigger ?? (
         <MenuButton
-          role="button"
```
Comment (0618) [inline packages/react/src/primitives/Menu/Menu.tsx:36]: not sure why we need to add it for a button, so removed it

File: packages/react/src/primitives/Rating/RatingIcon.tsx [hunk 6]
```diff
@@ -15,11 +15,9 @@ export const RatingIcon: React.FC<RatingIconProps> = ({
   className,
 }) => {
   return (
-    <View as="span" className={classNames(`amplify-rating-icon-container`)}>
-      <View as="label" className={classNames(`amplify-rating-label`)}>
```
Comment (0618) [inline packages/react/src/primitives/Rating/RatingIcon.tsx:19]: not sure what's the `label` for, so removed it

File: packages/react/src/primitives/Rating/RatingMixedIcon.tsx [hunk 7]
```diff
@@ -19,29 +19,29 @@ export const RatingMixedIcon: React.FC<RatingMixedIconProps> = ({
 }) => {
   const widthPercentage = `${(value % 1) * 100}%`;
   return (
-    <View as="span" className={classNames(`amplify-rating-icon-container`)}>
```
Comment (0618) [inline packages/react/src/primitives/Rating/RatingMixedIcon.tsx:22]: not sure why we need `classNames` here, so removed it

File: packages/ui/src/theme/css/component/rating.scss [hunk 8]
```diff
@@ -35,7 +35,7 @@
   width: 1em;
 }
 
-.amplify-rating-label {
+.amplify-rating-icon-place {
```
Comment (0618) [inline packages/ui/src/theme/css/component/rating.scss:38]: since the `label` tag is changed to `span`, rename the class

File: docs/src/pages/index.page.tsx [hunk 9]
```diff
@@ -133,7 +135,6 @@ const HomePage = ({ colorMode, setThemeOverride, themeOverride }) => {
       monoFont:
         '"Fira Mono", "DejaVu Sans Mono", Menlo, Consolas, "Liberation Mono", Monaco, "Lucida Console", monospace',
       fontSize: '16px',
-      lineHeight: '1.5',
```
Comment (0618) [inline docs/src/pages/index.page.tsx:136]: Remove it because of this error

Screen Shot 2022-03-04 at 14 45 55

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx [hunk 10]
```diff
@@ -100,6 +101,7 @@ export function SignIn() {
 
 SignIn.Header = (): JSX.Element => null;
 SignIn.Footer = () => {
+  // eslint-disable-next-line react-hooks/rules-of-hooks
```
Comment (reesscot) [inline packages/react/src/components/Authenticator/SignIn/SignIn.tsx:104]: Let's not disable rules of hooks here. There are a handful of rules of hooks violations in Authenticator which need to be dealt with and I don't want to have to find them manually.

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx [hunk 11]
```diff
@@ -61,6 +61,7 @@ export function SignIn() {
             className="amplify-flex"
             disabled={isPending}
           >
+            <legend>Sign in</legend>
```
Comment (reesscot) [inline packages/react/src/components/Authenticator/SignIn/SignIn.tsx:64]: Will this be translated? Also, this changes the design of the Authenticator. Should this be visually hidden @ErikCH @wlee221

File: docs/src/pages/index.page.tsx (hunk 9 above)
Comment (reesscot) [reply docs/src/pages/index.page.tsx:136]: Do we know why it's not present? Should we create a ticket to add it?

File: docs/src/pages/HomePrimitivePreview.tsx [hunk 12]
```diff
@@ -57,16 +57,16 @@ export const HomePrimitivePreview = () => {
           isExclusive
           onChange={(value: string) => setExclusiveValue(value)}
         >
-          <ToggleButton value="align-left">
+          <ToggleButton value="align-left" aria-label="align-left">
```
Comment (reesscot) [inline docs/src/pages/HomePrimitivePreview.tsx:60]:
```suggestion
          <ToggleButton value="align-left" ariaLabel="align-left">
```

File: docs/src/components/Layout/Header.tsx (hunk 3 above)
Comment (reesscot) [reply docs/src/components/Layout/Header.tsx:116]: Please use ariaLabel instead

File: packages/react/src/primitives/Rating/RatingMixedIcon.tsx [hunk 13]
```diff
@@ -19,29 +19,29 @@ export const RatingMixedIcon: React.FC<RatingMixedIconProps> = ({
...
           color={emptyColor}
         >
           {emptyIcon}
         </View>
       </View>
       <View
-        as="label"
-        className={classNames(`amplify-rating-label`)}
+        as="span"
+        className="amplify-rating-icon-place"
```
Comment (reesscot) [inline packages/react/src/primitives/Rating/RatingMixedIcon.tsx:37]: Please don't change this classname

**Review by reesscot — CHANGES_REQUESTED** [review]
Thanks for fixing these issues. Have some change requests below:

File: packages/react/src/primitives/Menu/Menu.tsx (hunk 5 above)
Comment (0618) [reply packages/react/src/primitives/Menu/Menu.tsx:36]: Nope, it doesn't cause any a11y error because it is a button. It would cause error if it's a `<div>` or something but uses as a button

File: packages/ui/src/theme/css/component/rating.scss (hunk 8 above)
[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/ui/src/theme/css/component/rating.scss:38]: Yes, making changes to class name is breaking changes.

File: docs/src/pages/index.page.tsx [hunk 14]
```diff
@@ -25,6 +25,8 @@ import { HomePrimitivePreview } from './HomePrimitivePreview';
 import { Copy } from '@/components/Copy';
 import { Footer } from '@/components/Layout/Footer';
 
+import type { SandpackThemeProp } from '@codesandbox/sandpack-react';
```
Comment (zchenwei) [inline docs/src/pages/index.page.tsx:28]: FYI, we are going to remove `sandpack` in favor of `react-live`. #1395  @dbanksdesign

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx (hunk 11 above)
Comment (zchenwei) [reply packages/react/src/components/Authenticator/SignIn/SignIn.tsx:64]: Yes this will change the design of Authenticator. It would be better if you can ping them to take a look at

File: packages/react/src/primitives/types/button.ts [hunk 15]
```diff
@@ -49,4 +49,9 @@ export interface ButtonProps extends ViewProps, FlexContainerStyleProps {
    * Changes the visual weight of the button.
    */
   variation?: ButtonVariations;
+
+  /**
+   * Changes href of the button.
+   */
+  href?: string;
```
[the remark this answers is not in the document]
Comment (dbanksdesign) [reply packages/react/src/primitives/types/button.ts:56]: We are using it in some places on the docs site where we do `<Button as={Link}` or `<Button as="a"`. I agree it is a bit weird to have href on the button type. I wonder if there is a better way to get a link that looks like a button?

File: packages/ui/src/theme/css/component/card.scss [hunk 16]
```diff
@@ -7,17 +7,18 @@
   box-shadow: var(--amplify-components-card-box-shadow);
   display: inline-block;
```
[the remark this answers is not in the document]
Comment (dbanksdesign) [reply packages/ui/src/theme/css/component/card.scss:8]: Makes sense to me.

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx (hunk 11 above)
Comment (dbanksdesign) [reply packages/react/src/components/Authenticator/SignIn/SignIn.tsx:64]: Does this fix an accessibility issue?

File: packages/react/src/primitives/TextField/TextField.tsx [hunk 17]
```diff
@@ -46,6 +46,8 @@ const TextFieldPrimitive = <Multiline extends boolean>(
 
   const fieldId = useStableId(id);
   const descriptionId = useStableId();
+  const labelId = useStableId();
+  const ariaDescribedBy = labelHidden ? labelId : descriptionId;
```
[the remark this answers is not in the document]
Comment (dbanksdesign) [reply packages/react/src/primitives/TextField/TextField.tsx:50]: I agree @reesscot's fixes

**Review by zchenwei — COMMENTED** [review]
I am just half way through it. Publishing a few comments first

File: packages/react/src/primitives/TextField/TextField.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/react/src/primitives/TextField/TextField.tsx:50]: Yup, agree! Thanks for the findings here and fixing this

File: packages/react/src/primitives/types/button.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/react/src/primitives/types/button.ts:56]: +1.. Should we just use an `a` tag in the case?

File: packages/ui/src/theme/css/component/card.scss (hunk 16 above)
[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/ui/src/theme/css/component/card.scss:8]: Changing this is a breaking change as well I think

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx (hunk 11 above)
Comment (0618) [reply packages/react/src/components/Authenticator/SignIn/SignIn.tsx:64]: Yes, there's an alert that `fieldset` missing legend

File: docs/src/pages/index.page.tsx (hunk 9 above)
Comment (0618) [reply docs/src/pages/index.page.tsx:136]: `SandpackThemeProp` doesn't have it so I think we can't add it? 

Also @zchenwei  [said](https://github.com/aws-amplify/amplify-ui/pull/1483#r821022178) we are removing Sandpack https://github.com/aws-amplify/amplify-ui/pull/1395 I think It might not matter

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx (hunk 11 above)
Comment (0618) [reply packages/react/src/components/Authenticator/SignIn/SignIn.tsx:64]: I forgot to add a `amplify-visually-hidden` class...I'll do that

File: packages/react/src/primitives/Menu/Menu.tsx (hunk 5 above)
Comment (0618) [reply packages/react/src/primitives/Menu/Menu.tsx:36]: I just found that our `<MenuButton>` is `button`, but `<Button>` is `div` 😱  

So we can remove `role` here, but we need to keep `role` for `<Button>`

File: packages/react/src/primitives/types/button.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (0618) [reply packages/react/src/primitives/types/button.ts:56]: Good point! I think we should not use buttons as anchor or vice versa. The best way is to make `button-link` and `link-button` CSS styles.

Also, we are using `div` for `button` and adding `role="button`, I think we may need to refactor that as well.

I'll remove the type change here and add a ticket for refactor.

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx (hunk 11 above)
Comment (wlee221) [reply packages/react/src/components/Authenticator/SignIn/SignIn.tsx:64]: Yep, adding visually-hidden should be sufficient.

File: docs/src/components/Layout/Header.tsx (hunk 3 above)
Comment (0618) [reply docs/src/components/Layout/Header.tsx]: Done.

File: packages/ui/src/theme/css/component/card.scss (hunk 16 above)
[the remark this answers is not in the document]
Comment (0618) [reply packages/ui/src/theme/css/component/card.scss]: Done. Changed to `block`

File: packages/ui/src/theme/css/component/rating.scss (hunk 8 above)
[the remark this answers is not in the document]
Comment (0618) [reply packages/ui/src/theme/css/component/rating.scss]: Reverted back.

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx [hunk 18]
```diff
@@ -61,7 +61,7 @@ export function SignIn() {
             className="amplify-flex"
             disabled={isPending}
           >
-            <legend>Sign in</legend>
+            <legend className="amplify-visually-hidden">Sign in</legend>
```
Comment (reesscot) [inline packages/react/src/components/Authenticator/SignIn/SignIn.tsx:64]: Instead of using the classname, let's wrap this in the `VisuallyHidden` component instead for consistency

File: packages/react/src/primitives/Rating/RatingIcon.tsx (hunk 6 above)
Comment (jacoblogan) [reply packages/react/src/primitives/Rating/RatingIcon.tsx:19]: I believe the label tag was a remnant of earlier designs when this component was not read only.  Looking at some other rating components out there they use a label and hidden checkbox to mark user selections.  @reesscot brought up a good point that we could have customers who are targeting this label for styling and so if we don't need to remove it, then it would be best to leave it in.  @0618 other than the extra clutter, was the `label` causing any problems? If not then I agree that we should leave it in as to not create a potentially breaking change for customers.

File: packages/react/src/primitives/TextField/TextField.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (0618) [reply packages/react/src/primitives/TextField/TextField.tsx]: Updated! Converted the change and hide `FieldDescription` if `labelHidden` is `true`, instead of `FieldDescription` returning null.

File: packages/react/src/primitives/Rating/RatingIcon.tsx (hunk 6 above)
Comment (0618) [reply packages/react/src/primitives/Rating/RatingIcon.tsx:19]: @jacoblogan yes, "empty label" is a type of a11y error, so I removed it but kept the class selector in this PR. Is that ok?

File: packages/react/src/primitives/Rating/RatingMixedIcon.tsx (hunk 13 above)
Comment (0618) [reply packages/react/src/primitives/Rating/RatingMixedIcon.tsx]: done

File: packages/react/src/primitives/Menu/Menu.tsx [hunk 19]
```diff
@@ -33,7 +33,7 @@ const MenuPrimitive: Primitive<MenuProps, 'div'> = (
...
+          ariaLabel="menu button"
```
[the remark this answers is not in the document]
Comment (0618) [reply packages/react/src/primitives/Menu/Menu.tsx]: Good point! Done

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx (hunk 18 above)
Comment (0618) [reply packages/react/src/components/Authenticator/SignIn/SignIn.tsx]: Done

File: docs/src/pages/index.page.tsx (hunk 14 above)
Comment (dbanksdesign) [reply docs/src/pages/index.page.tsx:28]: potentially... there are some potential security issues with it since it uses `eval`, still up for debate

File: packages/react/src/primitives/Field/FieldDescription.tsx [hunk 20]
```diff
@@ -8,12 +9,15 @@ export const FieldDescription: React.FC<FieldDescriptionProps> = ({
...
-  return !labelHidden && descriptiveText ? (
-    <Text className={ComponentClassNames.FieldDescription} {...rest}>
-      {descriptiveText}
-    </Text>
-  ) : null;
-};
+}) => (
+  <Text
+    className={classNames(ComponentClassNames.FieldDescription, {
+      'amplify-visually-hidden': labelHidden,
```
Comment (dbanksdesign) [inline packages/react/src/primitives/Field/FieldDescription.tsx:15]: This could be `ComponentClassNames.VisuallyHidden`

**Review by dbanksdesign — APPROVED** [review]
LGTM! Just one non-blocking comment

File: .changeset/quiet-impalas-chew.md [hunk 21]
```diff
@@ -0,0 +1,8 @@
+---
+'@aws-amplify/ui-react': patch
```
Comment (jacoblogan) [inline .changeset/quiet-impalas-chew.md:2]: with the new attributes being added and the structure of some of the components being updated do we want this to be a minor bump instead of a patch?

Comment (0618) [reply .changeset/quiet-impalas-chew.md:2]: Good point...

I checked the [semver doc](https://semver.org/#spec-item-7)

> Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backwards compatible functionality is introduced to the public API. It MUST be incremented if any public API functionality is marked as deprecated. It MAY be incremented if substantial new functionality or improvements are introduced within the private code. It MAY include patch level changes. Patch version MUST be reset to 0 when minor version is incremented.

Since we don't have any new features in this PR, I think it's not a MUST.

File: packages/react/src/primitives/Field/FieldDescription.tsx (hunk 20 above)
Comment (0618) [reply packages/react/src/primitives/Field/FieldDescription.tsx]: Done

**Review by jacoblogan — APPROVED** [review]
LGTM

File: packages/react/src/components/Authenticator/SignIn/SignIn.tsx [hunk 22]
```diff
@@ -61,6 +61,9 @@ export function SignIn() {
             className="amplify-flex"
             disabled={isPending}
           >
+            <VisuallyHidden>
+              <legend>Sign in</legend>
```
Comment (reesscot) [inline packages/react/src/components/Authenticator/SignIn/SignIn.tsx:65]: Please split this change out into a separate PR and include the Vue and Angular SignIn screens as well. I'd like this to be reviewed in a non-docs change PR since it's more about Authenticator accessibility than strictly the docs.

File: packages/react/src/primitives/SliderField/SliderField.tsx [hunk 23]
```diff
@@ -50,6 +50,7 @@ const SliderFieldPrimitive: Primitive<SliderFieldProps, typeof Root> = (
   const fieldId = useStableId(id);
   const labelId = useStableId();
   const descriptionId = useStableId();
+  const ariaDescribedBy = labelHidden ? labelId : descriptionId;
```
[the remark this answers is not in the document]
Comment (reesscot) [reply packages/react/src/primitives/SliderField/SliderField.tsx:53]: Please also add a unit test for this.

[the remark this answers is not in the document]
Comment (0618) [reply packages/react/src/primitives/SliderField/SliderField.tsx:53]: Weird...I removed the code in this commit https://github.com/aws-amplify/amplify-ui/pull/1483/commits/94bb127d1a531dfb87b83b71cbee055d73d6f1ae but it still shows up in the latest PR...

File: packages/react/src/primitives/TextField/TextField.tsx [hunk 24]
```diff
@@ -98,7 +99,7 @@ const TextFieldPrimitive = <Multiline extends boolean>(
       testId={testId}
       {...flexContainerStyleProps}
     >
-      <Label htmlFor={fieldId} visuallyHidden={labelHidden}>
+      <Label htmlFor={fieldId} visuallyHidden={labelHidden} id={labelId}>
```
Comment (reesscot) [inline packages/react/src/primitives/TextField/TextField.tsx:102]: Why do we need this? the `htmlFor` is handling the link to the `input` field.

File: packages/react/src/primitives/Rating/RatingIcon.tsx (hunk 6 above)
Comment (reesscot) [reply packages/react/src/primitives/Rating/RatingIcon.tsx:19]: Makes sense. If it's causing a11y error let's remove it.

File: packages/react/src/primitives/Menu/Menu.tsx (hunk 5 above)
Comment (reesscot) [reply packages/react/src/primitives/Menu/Menu.tsx:36]: Tested it and the menu button still works correctly. Proceed 👍

File: packages/react/src/primitives/SliderField/SliderField.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (reesscot) [reply packages/react/src/primitives/SliderField/SliderField.tsx:53]: Actually, can we please split these changes out into a separate PR, as we will need to make these changes in all the Field components that use the `FieldDescription` component.

**Review by reesscot — CHANGES_REQUESTED** [review]

File: packages/react/src/primitives/Field/FieldDescription.tsx [hunk 25]
```diff
@@ -1,4 +1,5 @@
 import * as React from 'react';
+import classNames from 'classnames';
```
Comment (reesscot) [inline packages/react/src/primitives/Field/FieldDescription.tsx:2]: remove unused import

**Review by reesscot — APPROVED** [review]
LGTM, just one small change (removing a now unused import)
